File: flow-server/src/main/resources/META-INF/frontend/WebPushToggle.ts
/*
 * Copyright 2000-2023 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
import {css, html, LitElement, nothing} from 'lit';
import {customElement, property, state} from 'lit/decorators.js';
import '@vaadin/button';
import '@vaadin/checkbox';
import {CheckboxCheckedChangedEvent} from "@vaadin/checkbox";

@customElement('web-push-toggle')
export class WebPushToggle extends LitElement {
  @property({type: String}) caption = 'Subscribe to push notifications';
  @property({type: String}) publicKey = '';
  @state() denied = Notification.permission === 'denied';
  @state() subscribed = false;

  static styles = css`
    :host {
      display: block;
    }
  `;

  render() {
    return html`
        ${this.denied
                ? html` <b> You have blocked notifications. You need to manually enable them in your browser. </b> `
                : nothing}
        <vaadin-checkbox
                ?checked=${this.subscribed}
                label=${this.caption}
                @checked-changed=${this.updateSubscription}
        ></vaadin-checkbox>

    `;
  }

  first = true;
  updateSubscription(e: CheckboxCheckedChangedEvent){

    // The checkbox fires an event on initialization, ignore it.
    if(this.first){
      this.first = false;
      return;
    }

    // The checkbox value is set on initialization based on the service worker subscription.
    // Don't re-subscribe if we're already subscribed.
    if(e.detail.value === this.subscribed) {
      return;
    }
    if(e.detail.value){
      this.subscribe();
    } else {
      this.unsubscribe();
    }
  }

  async connectedCallback() {
    super.connectedCallback();
    const registration = await navigator.serviceWorker.getRegistration();
    this.subscribed = !!(await registration?.pushManager.getSubscription());
  }

  async subscribe() {
    const notificationPermission = await Notification.requestPermission();

    if (notificationPermission === 'granted') {
      const registration = await navigator.serviceWorker.getRegistration();
      const subscription = await registration?.pushManager.subscribe({
        userVisibleOnly: true,
        applicationServerKey: this.urlB64ToUint8Array(this.publicKey),
      });

      if (subscription) {
        this.subscribed = true;
        console.log(subscription);
        console.log(JSON.parse(JSON.stringify(subscription)));
        this.dispatchEvent(new CustomEvent('web-push-subscribed', {
          bubbles: true,
          composed: true,
          // Serialize keys uint8array -> base64
          detail: JSON.parse(JSON.stringify(subscription))
        }));
      }
    } else {
      this.denied = true;
    }
  }

  async unsubscribe() {
    const registration = await navigator.serviceWorker.getRegistration();
    const subscription = await registration?.pushManager.getSubscription();
    if (subscription) {
      await subscription.unsubscribe();

      this.dispatchEvent(new CustomEvent('web-push-unsubscribed', {
        bubbles: true,
        composed: true,
        // Serialize keys uint8array -> base64
        detail: JSON.parse(JSON.stringify(subscription))
      }));

      this.subscribed = false;
    }
  }

  private urlB64ToUint8Array(base64String: string) {
    const padding = '='.repeat((4 - (base64String.length % 4)) % 4);
    const base64 = (base64String + padding).replace(/\-/g, '+').replace(/_/g, '/');
    const rawData = window.atob(base64);
    const outputArray = new Uint8Array(rawData.length);
    for (let i = 0; i < rawData.length; ++i) {
      outputArray[i] = rawData.charCodeAt(i);
    }
    return outputArray;
  }
}

File: flow-server/src/main/java/com/vaadin/flow/server/frontend/webpush/WebPushToggle.java
/*
 * Copyright 2000-2023 Vaadin Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package com.vaadin.flow.server.frontend.webpush;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEvent;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.DomEvent;
import com.vaadin.flow.component.EventData;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;

@Tag("web-push-toggle")
@JsModule("./WebPushToggle.ts")
public class WebPushToggle extends Component {

    public WebPushToggle(String publicKey) {
        setPublicKey(publicKey);
    }

    public void setPublicKey(String publicKey) {
        getElement().setProperty("publicKey", publicKey);
    }

    public void setCaption(String caption) {
        getElement().setProperty("caption", caption);
    }

    // Events

    public static class WebPushSubscriptionEvent
            extends ComponentEvent<WebPushToggle> {

        String endpoint;
        String auth;
        String p256dh;

        public WebPushSubscriptionEvent(WebPushToggle source,
                boolean fromClient,  String endpoint,
        String auth,
        String p256dh) {
            super(source, fromClient);
            this.endpoint = endpoint;
            this.auth = auth;
            this.p256dh = p256dh;
        }

        public String getEndpoint() {
            return endpoint;
        }

        public String getAuth() {
            return auth;
        }

        public String getP256dh() {
            return p256dh;
        }
    }

    @DomEvent("web-push-subscribed")
    public static class SubscribeEvent extends WebPushSubscriptionEvent {
        public SubscribeEvent(WebPushToggle source, boolean fromClient,
                @EventData("event.detail.endpoint") String endpoint,
                @EventData("event.detail.keys.auth") String auth,
                @EventData("event.detail.keys.p256dh") String p256dh) {
            super(source, fromClient, endpoint, p256dh, auth);
        }
    }

    @DomEvent("web-push-unsubscribed")
    public static class UnsubscribeEvent extends WebPushSubscriptionEvent {

        public UnsubscribeEvent(WebPushToggle source, boolean fromClient,
                @EventData("event.detail.endpoint") String endpoint,
                @EventData("event.detail.keys.auth") String auth,
                @EventData("event.detail.keys.p256dh") String p256dh) {

            super(source, fromClient, endpoint, p256dh, auth);
        }
    }

    public Registration addSubscribeListener(
            ComponentEventListener<SubscribeEvent> listener) {
        return addListener(SubscribeEvent.class, listener);
    }

    public Registration addUnsubscribeListener(
            ComponentEventListener<UnsubscribeEvent> listener) {
        return addListener(UnsubscribeEvent.class, listener);
    }
}
